spyder/plugins/workingdirectory/plugin.py

```diff
@@ -4,232 +4,165 @@
-        if osp.isfile(self.LOG_PATH):
-            wdhistory, _ = encoding.readlines(self.LOG_PATH)
-            wdhistory = [name for name in wdhistory if os.path.isdir(name)]
+
+    def get_description(self):
+        return _('Set the current working directory for various plugins.')
+
+    def get_icon(self):
+        return self.create_icon('DirOpenIcon')
+
+    def register(self):
+        container = self.get_container()
+        editor = self.get_plugin(Plugins.Editor)
+        explorer = self.get_plugin(Plugins.Explorer)
+        ipyconsole = self.get_plugin(Plugins.IPythonConsole)
+
+        self.add_application_toolbar(self.NAME, container.toolbar)
+        container.sig_current_directory_changed.connect(
+            self.sig_current_directory_changed)
+        self.sig_current_directory_changed.connect(
+            lambda path, plugin=None: self.chdir(path, plugin))
+        container.set_history(self.load_history())
+
+        if editor:
+            editor.sig_dir_opened.connect(
+                lambda path, plugin=editor: self.chdir(path, editor))
+
+        if ipyconsole:
+            self.sig_current_directory_changed.connect(
+                ipyconsole.set_current_client_working_directory)
+            # TODO: chdir_current_client might follow a better naming
+            # convention
+            ipyconsole.sig_current_directory_changed.connect(
+                lambda path, plugin=ipyconsole: self.chdir(path, plugin))
+
+        if explorer:
+            self.sig_current_directory_changed.connect(
+                lambda path: explorer.chdir(path, emit=False))
+            explorer.sig_dir_opened.connect(
+                lambda path, plugin=explorer: self.chdir(path, plugin))
+
+    # --- Public API
+    # ------------------------------------------------------------------------
+    def chdir(self, directory, sender_plugin=None):
+        """
+        Change current working directory.
+
+        Parameters
+        ----------
+        directory: str
+            The new working directory to set.
+        sender_plugin: spyder.api.plugins.SpyderPluginsV2
+            The plugin that requested this change: Default is None.
+        """
+        explorer = self.get_plugin(Plugins.Explorer)
+        ipyconsole = self.get_plugin(Plugins.IPythonConsole)
+
+        if explorer and sender_plugin != explorer:
+            explorer.chdir(directory, emit=False)
+            explorer.refresh_plugin(directory, force_current=True)
+
+        if ipyconsole and sender_plugin != ipyconsole:
+            ipyconsole.set_current_client_working_directory(directory)
+
+        if sender_plugin is not None:
+            container = self.get_container()
+            container.chdir(directory, emit=False)
+
+        self.save_history()
+
+    def load_history(self, workdir=None):
+        """
+        Load history from a text file located in Spyder configuration
+        folder and sets the `workdir`.

```

Comment:
As far as I can tell from the code below, this method doesn't set the working directory. Instead, it just returns the value of `get_workdir` if there are no directories previously saved in `LOG_PATH`.

Reply:
You are right, it only adds it to the history, but does not set it. Maybe it should set the first entry of the history as the CWD?

Reply:
Well, until there's a new directory selected in the toolbar, nothing is going to be saved to LOG_PATH. So I think the logic is fine for now, although it could be reviewed in another PR.

Reply:
Ok